src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb

```diff
@@ -1089,6 +1099,102 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
         ''' </summary>
         Friend MustOverride Function GetSynthesizedWithEventsOverrides() As IEnumerable(Of PropertySymbol)
 
+        Friend ReadOnly Property AllRequiredMembers As ImmutableSegmentedDictionary(Of String, Symbol)
+            Get
+                TryCalculateRequiredMembersWithCache(Nothing)
+                Debug.Assert(Not _lazyRequiredMembers.IsDefault)
+                Return _lazyRequiredMembers
+            End Get
+        End Property
+
+        Friend ReadOnly Property HasRequiredMembersError As Boolean
+            Get
+                TryCalculateRequiredMembersWithCache(Nothing)
+                Debug.Assert(Not _lazyRequiredMembers.IsDefault)
+                Return _lazyRequiredMembers = s_requiredMembersErrorSentinel
+            End Get
+        End Property
+
+        Friend MustOverride ReadOnly Property HasAnyDeclaredRequiredMembers As Boolean
+
+        Private Function TryCalculateRequiredMembersWithCache(<Out> ByRef requiredMembersBuilder As ImmutableSegmentedDictionary(Of String, Symbol).Builder) As Boolean
+            Debug.Assert(requiredMembersBuilder Is Nothing)
+            Dim lazyRequiredMembers As ImmutableSegmentedDictionary(Of String, Symbol) = _lazyRequiredMembers
+
+            If Not lazyRequiredMembers.IsDefault Then
+                If lazyRequiredMembers = s_requiredMembersErrorSentinel Then
+                    Return False
+                Else
+                    Return True
+                End If
+            End If
+
+            Dim success = TryCalculateRequiredMembers(requiredMembersBuilder)
+
+            Dim requiredMembers = If(success,
+                If(requiredMembersBuilder?.ToImmutable(), If(BaseTypeNoUseSiteDiagnostics?.AllRequiredMembers, ImmutableSegmentedDictionary(Of String, Symbol).Empty)),
+                s_requiredMembersErrorSentinel)
+
+            RoslynImmutableInterlocked.InterlockedInitialize(_lazyRequiredMembers, requiredMembers)
+
+            Return success
+        End Function
+
+        Private Function TryCalculateRequiredMembers(<[In], Out> ByRef requiredMembersBuilder As ImmutableSegmentedDictionary(Of String, Symbol).Builder) As Boolean
```

Comment:
><[In], Out> ByRef requiredMembersBuilder As ImmutableSegmentedDictionary(Of String, Symbol).Builder [](http://example.com/codeflow?start=53&length=99)

I think that "dragging" the builder along add a fair amount  of complexity. Is this a case of premature optimization, or was this added to address a real observable problem?

Reply:
I don't feel that it adds a significant amount of complexity. The base calls will still need to occur, we'll still need to have the dual builder of members for the current type and dictionary of members in the base type that we have below. It would remove a parameter from this method and replace a call to `TryCalculateRequiredMembers` with `HasRequiredMembersError`, but the logic would not be significantly affected.

Reply:
I am not talking about computational complexity, but the complexity of the logic. In my opinion the current logic is more complex than it could be. Unnecessary complexity is a source of bugs. So far I haven't heard what advantage do we get for taking this risk.